<a href="https://colab.research.google.com/github/egitonga395/kamilimu_assignments/blob/erick_python_libraries_1/Erick_data_manipulation_with_python_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Question 1
1. Helps with dealing with duplicate data
2. Missing data
3. Removing duplicates
4. Correcting the data type fomarts

# Question 2
1. Pyspark
2. Apache
3. Numpy

# Question 3
1. Numpy is very fast as it with numerical data while compared to lists
2. Numpy's syntax is allowed for mathematical operations without the need of for loops which adds on to the complexity and utiliztion of more computing resources

# Question 4
1. While list concantete when mathematical operations are introcude, Numpy allows for direct mathematical operations
2. It is difficult to visualize the arrays in list as it involves nesting  of 2 lists numpy on the other hands presents the data as columns and rows

# Question 5
An Ndarray refers to Nth dimensional Array, This refers to data dat which is orgganized in an array, which could be nested in each other. This data is often used to give representation of real world models, for example distance is a one dimension, speed is two dimension[matrix], volume is three dimension, space-time as 4d  

#Question 6
1. Numpy is faster than list interms of computation
2. Numpy is much efficient interms of computer resources
3. Numpy is much easier to use
4. Numpy is much more intuitive, especially when use with pandas model the data visually

# Question 7
1. Series is used with one dimensional data
2. Dataframe is used to deal with two dimensional data such as spreadsheets and datasets/tables from databases

#Question 8
Pandas arranges the data into rows and columns making the data easier to use
futhermore the data can be indexed to ensure that it is intuitive

#Question 9
df["column_name"]
df.column_name
supposed your dataframe has a colums called order_id that would like to be accessed
df["order_id"]
or
df.order_id
can be used
# Question 10

df.isnull().sum() is used to check for the total NaN values in a dataset being analysed. It is important to so so as to find out whether some row values will be dropped or repopulated. This is done inorder to avoid errors during mathematical computations. Null values often cause errors during analysis since they are the absence of a value.

Loading the sheets

In [42]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep='\t')

Look at the data first


In [43]:
df.tail(7)

order_id  quantity            item_name  \
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                NaN     $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

Look at the description of the data

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


There are 4622 counts.  Choice descrition has NaNs.

How many entries item_price is none in the dataset

In [45]:
df["item_price"].isnull().sum()

np.int64(0)

What are the columns names in the dataset

In [46]:
print(df.columns)

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')


Identify the item name that appears most frequestly

In [47]:

df.item_name.mode()

0    Chicken Bowl
Name: item_name, dtype: object

how many distict entries are here? non-duplicated data. Item name does not contaun any NaN values

In [48]:
print(df.item_name.count() - df.duplicated().sum())

4563


creating a new dataframe which only has the name chickenbowl

In [49]:
df2 = df[df['item_name'] == 'Chicken Bowl']
df2.head()

order_id  quantity     item_name  \
4          2         2  Chicken Bowl   
5          3         1  Chicken Bowl   
13         7         1  Chicken Bowl   
19        10         1  Chicken Bowl   
26        13         1  Chicken Bowl   

                                   choice_description item_price  
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
13  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25   
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...     $8.75   
26  [Roasted Chili Corn Salsa (Medium), [Pinto Bea...     $8.49

Remove the dollar sign
Convert it to a float type.

In [55]:
df["item_price"] = df["item_price"].str.lstrip('$')
df["item_price"] = df["item_price"].astype(float)
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

average price acrosss all orders


In [56]:
df.item_price.mean()

np.float64(7.464335785374297)

Count no of unique orders

In [57]:
df3 = df.drop_duplicates(subset=['order_id'])
df3.order_id.count()

np.int64(1834)

How calculate grand total

In [58]:
df["garnd_total"] = df["quantity"] * df["item_price"]
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  garnd_total  
0                                                NaN        2.39         2.39  
1                                       [Clementine]        3.39         3.39  
2                                            [Apple]        3.39         3.39  
3                                                NaN        2.39         2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98        33.96